crt/fork.c

```diff
@@ -0,0 +1,569 @@
+#include <myst/lfence.h>
+#include <myst/round.h>
+#include <myst/setjmp.h>
+#include <myst/syscallext.h>
+#include <myst/thread.h>
+
+#define DEFAULT_STACK_SIZE (4 * 1024 * 1024)
+
+#define DEFAULT_GUARD_SIZE 8192
+
+#define hidden __attribute__((__visibility__("hidden")))
+
+struct __libc
+{
+    int can_do_threads;
+    int threaded;
+    int secure;
+    volatile int threads_minus_1;
+    size_t* auxv;
+    struct tls_module* tls_head;
+    size_t tls_size;
+#if 0
+    size_t tls_align;
+    size_t tls_cnt;
+    size_t page_size;
+    struct __locale_struct global_locale;
+#endif
+};
+
+extern hidden struct __libc __libc;
+
+enum
+{
+    DT_EXITING = 0,
+    DT_JOINABLE,
+    DT_DETACHED,
+};
+
+struct pthread
+{
+    struct pthread* self;
+    uintptr_t* dtv;
+    struct pthread* prev;
+    struct pthread* next;
+    uintptr_t sysinfo;
+    uintptr_t canary;
+    uintptr_t canary2;
+    int tid;
+    int errno_val;
+    volatile int detach_state;
+    volatile int cancel;
+    volatile unsigned char canceldisable;
+    volatile unsigned char cancelasync;
+    unsigned char tsd_used : 1;
+    unsigned char dlerror_flag : 1;
+    unsigned char* map_base;
+    size_t map_size;
+    void* stack;
+    size_t stack_size;
+    size_t guard_size;
+    void* result;
+    struct __ptcb* cancelbuf;
+    void** tsd;
+    struct
+    {
+        volatile void* volatile head;
+        long off;
+        volatile void* volatile pending;
+    } robust_list;
+    volatile int timer_id;
+    locale_t locale;
+    volatile int killlock[1];
+    char* dlerror_buf;
+    void* stdio_locks;
+    uintptr_t canary_at_end;
+    uintptr_t* dtv_copy;
+};
+
+int myst_round_up(uint64_t x, uint64_t m, uint64_t* r)
+{
+    uint64_t t;
+
+    if (!r)
+        return -EINVAL;
+
+    /* prevent divide by zero */
+    if (m == 0)
+        return -ERANGE;
+
+    if (__builtin_add_overflow(x, m - 1, &t))
+        return -ERANGE;
+
+    if (__builtin_mul_overflow(t / m, m, r))
+        return -ERANGE;
+
+    return 0;
+}
+
+/* ATTN: does not work for Linux yet: SYS_arch_prctl */
+static void _set_fsbase(void* p)
+{
+    __asm__ volatile("wrfsbase %0" ::"r"(p));
+}
+
+static inline struct pthread* __pthread_self()
+{
+    struct pthread* self;
+    __asm__("mov %%fs:0,%0" : "=r"(self));
+    return self;
+}
+
+extern hidden unsigned __default_guardsize;
+extern hidden unsigned __default_stacksize;
+extern hidden volatile size_t __pthread_tsd_size;
+
+extern hidden void* __copy_tls(unsigned char*);
+
+static void _get_parent_stack(void** stack, size_t* stack_size)
+{
+    struct pthread* self = __pthread_self();
+
+    if (self->stack && self->stack_size)
+    {
+        *stack = (uint8_t*)self->stack - self->stack_size;
+        *stack_size = self->stack_size;
+    }
+    else
+    {
+        const long n = SYS_myst_get_process_stack;
+
+        if (syscall(n, stack, stack_size) != 0)
+        {
+            fprintf(stderr, "cannot retrieve parent stack\n");
```

Comment:
Do you mean "the process thread of the parent?" My understanding is that the thread that calls "fork" is the parent thread, but it is not necessarily the process thread.

Reply:
hmm... The only thread we cannot get the stack from pthread is the main process thread. All other threads are created through pthread_create. 
If this is called from a non-main-process-thread the pthread stack is sufficient, but it is a little confusing. 
I will change the name, and will fail if not called on the main process thread as if we were to return the main process stack from a non-main process thread it would cause all sorts of problems.